tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java

```diff
@@ -0,0 +1,1500 @@
+        Map<TopicPartitionReplica, String> targetReplicas = t0.v2;
+        Set<TopicPartition> targetPartsSet = targetParts.stream().map(t -> t.v1).collect(Collectors.toSet());
+        Set<TopicPartition> curReassigningParts = new HashSet<>();
+        adminClient.listPartitionReassignments(targetPartsSet).reassignments().get().forEach((part, reassignment) -> {
+            if (reassignment.addingReplicas().isEmpty() || !reassignment.removingReplicas().isEmpty())
+                curReassigningParts.add(part);
+        });
+        if (!curReassigningParts.isEmpty()) {
+            Map<TopicPartition, Throwable> errors = cancelPartitionReassignments(adminClient, curReassigningParts);
+            if (!errors.isEmpty()) {
+                throw new TerseException(String.format(
+                    "Error cancelling partition reassignment%s for:%n%s",
+                    errors.size() == 1 ? "" : "s",
+                    errors.keySet().stream()
+                        .sorted(ReassignPartitionsCommand::compareTopicPartitions)
+                        .map(part -> part + ": " + errors.get(part).getMessage()).collect(Collectors.joining(System.lineSeparator())))
+                );
+            }
+            System.out.printf("Successfully cancelled partition reassignment%s for: %s%n",
+                curReassigningParts.size() == 1 ? "" : "s",
+                curReassigningParts.stream().sorted(ReassignPartitionsCommand::compareTopicPartitions).map(Object::toString).collect(Collectors.joining(","))
+            );
+        } else {
+            System.out.println("None of the specified partition reassignments are active.");
+        }
+        Map<TopicPartitionReplica, String> curMovingParts = new HashMap<>();
+        findLogDirMoveStates(adminClient, targetReplicas).forEach((part, moveState) -> {
+            if (moveState instanceof ActiveMoveState)
+                curMovingParts.put(part, ((ActiveMoveState) moveState).currentLogDir);
+        });
+        if (curMovingParts.isEmpty()) {
+            System.out.print("None of the specified partition moves are active.");
+        } else {
+            executeMoves(adminClient, curMovingParts, timeoutMs, time);
+        }
+        if (!preserveThrottles) {
+            clearAllThrottles(adminClient, targetParts);
+        }
+        return new Tuple2<>(curReassigningParts, curMovingParts.keySet());
+    }
+
+    private static String formatAsReassignmentJson(Map<TopicPartition, List<Integer>> partitionsToBeReassigned,
+                                                   Map<TopicPartitionReplica, String> replicaLogDirAssignment) throws JsonProcessingException {
+        List<Map<String, Object>> partitions = new ArrayList<>();
+        partitionsToBeReassigned.keySet().stream().sorted(ReassignPartitionsCommand::compareTopicPartitions).forEach(tp -> {
+            List<Integer> replicas = partitionsToBeReassigned.get(tp);
+            Map<String, Object> data = new LinkedHashMap<>();
```

Comment:
Why `LinkedHashMap`? Is there any contract on the order of the fields? Does scala `Map` with `.asJAva` guarantee  keys order?

Reply:
> Does scala Map with .asJAva guarantee keys order?

Yes. Please, take a look to `ReassignPartitionsUnitTest#testCurrentPartitionReplicaAssignmentToString`